src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala

```diff
@@ -51,38 +57,81 @@ class HttpGoogleServicesDAO(
   val jsonFactory = JacksonFactory.getDefaultInstance
   val clientSecrets = GoogleClientSecrets.load(jsonFactory, new StringReader(clientSecretsJson))
 
-  val system = akka.actor.ActorSystem("system")
-
-  override def createBucket(userInfo: UserInfo, projectId: String, workspaceId: String, workspaceName: WorkspaceName): Unit = {
+  override def createBucket(userInfo: UserInfo, projectId: String, workspaceId: String, workspaceName: WorkspaceName): Future[Unit] = {
     val bucketName = getBucketName(workspaceId)
     val directory = getGroupDirectory
     val groups = directory.groups
-    insertGroup(groups,newGroup(bucketName,workspaceName,WorkspaceAccessLevel.Read)) {
-      insertGroup(groups,newGroup(bucketName,workspaceName,WorkspaceAccessLevel.Write)) {
-        insertGroup(groups,newGroup(bucketName,workspaceName,WorkspaceAccessLevel.Owner)) {
-          val ownersGroupId = toGroupId(bucketName,WorkspaceAccessLevel.Owner)
+
+    // tries to delete all groups expected to have been created whether they actually were or not
+    def rollbackGroups(t: Throwable): Throwable = {
+      Future.traverse(groupAccessLevelsAscending) { accessLevel =>
+        Future {
+          val inserter = groups.delete(newGroup(bucketName, workspaceName, accessLevel).getEmail)
+          blocking {
+            inserter.execute
+          }
+        }
+      }
+      t
+    }
+
+    def insertGroups(workspaceName: WorkspaceName, bucketName: String): Future[Seq[Try[Group]]] = {
+      Future.traverse(groupAccessLevelsAscending) { accessLevel =>
+        toFutureTry(retry(when500) {
+          () => Future {
+            val inserter = groups.insert(newGroup(bucketName, workspaceName, accessLevel))
+            blocking {
+              inserter.execute
+            }
+          }
+        })
+      }
+    }
+
+    def assertSuccessfulGroupInserts: (Seq[Try[Group]]) => Future[Seq[Group]] = { tryGroups =>
```

Comment:
Obscure as written.  Consider returning a more intuitive type.  The only real criterion on the return type is that it's a monad so that you can use it in your flatMapping pipeline.  Or break the pipeline in two and just use a simple predicate here.